server/src/main/java/org/elasticsearch/repositories/Repository.java

```diff
@@ -117,13 +118,17 @@ public void onFailure(Exception e) {
     IndexMetadata getSnapshotIndexMetaData(RepositoryData repositoryData, SnapshotId snapshotId, IndexId index) throws IOException;
 
     /**
-     * Returns a {@link RepositoryData} to describe the data in the repository, including the snapshots
-     * and the indices across all snapshots found in the repository.  Throws a {@link RepositoryException}
-     * if there was an error in reading the data.
-     * @param listener listener that may be resolved on different kinds of threads including transport and cluster state applier threads
-     *                 and therefore must fork to a new thread for executing any long running actions
+     * Returns a {@link RepositoryData} to describe the data in the repository, including the snapshots and the indices across all snapshots
+     * found in the repository. Completes the listener with a {@link RepositoryException} if there was an error in reading the data.
+     *
+     * @param responseExecutor Executor to use to complete the listener if not using the calling thread. Using {@link
+     *                         org.elasticsearch.common.util.concurrent.EsExecutors#DIRECT_EXECUTOR_SERVICE} means to complete the listener
+     *                         on the thread which ultimately resolved the {@link RepositoryData}, which might be a low-latency transport or
+     *                         cluster applier thread so make sure not to do anything slow or expensive in that case.
+     * @param listener         Listener which is either completed on the calling thread (if the {@link RepositoryData} is immediately
+     *                         available, e.g. from an in-memory cache), otherwise it is completed using {@code responseExecutor}.
      */
-    void getRepositoryData(ActionListener<RepositoryData> listener);
+    void getRepositoryData(Executor responseExecutor, ActionListener<RepositoryData> listener);
```

Comment:
I wonder whether it is worthwhile to abstract this pattern to similar to `ThreadedActionListener`. But instead of alway completing the listener with the executor like `ThreadedActionListener` does, only use the executor when the thread is different from the one that creates the listener. 

```java
class MaybeThreadedActionListener extends ThreadedActionListener {
    public MaybeThreadedActionListener(Executor executor, ActionListener delegate) {
        this.creationThread = Thread.currentThread();
        super(executor, delegate);
    }

    public void onResponse(Response response) {
        if (Thread.currentThread() == creationThread) {
            delegate.onResponse(response);
        } else {
            super.onResponse(response);
        }
    }
    ...
}
```
I am aware that `ThreadedActionListener` is final. The above code is just to demostrate the idea, not an actual working version. 

This seems to reduce the downstream complexity of setting the listener correctly depending on how repo data is obtained. What do you think?

Reply:
I know what you mean, but I think `MaybeThreadedActionListener` might be a little trappy too. We could fork, and then fork back and end up on the original thread again purely by coincidence, and in that situation I don't think we do always want to continue running on that thread. For instance, a single thread could run hundreds of tasks which fork and all end up waiting on the same listener. Then if that listener eventually happens to complete on the original thread, we'd run all those tasks sequentially in a loop rather than parallelising them across the target threadpool as the caller might expect.

Reply:
Not sure if I fully understand. Are you saying that this pattern is trappy to abstract because being on the same thread when listener completes does not mean the work has not forked previously? So instead we should manually ensure each implementation of `getRepositoryData` use the right thread to complete?

Reply:
The pattern (as presented in this PR) is fine IMO, but yes checking for thread equality is not a reliable way to be sure if more forking is needed.

Reply:
Fair enought. Thanks for the discussion.